### ISSUE TITLE
Unable to build macos desktop: error during CC dnabasic.lo

### ISSUE THREAD

**Issue by JasonInOttawa** [koreader/koreader#6687]
* KOReader version: 2020.09
* Device: macOS 10.15.6, XCode 12

#### Issue

Error during run of ./kodev release macos:

Making install in src
...
  CC       dnabasic.lo
dewarp2.c:144:8: error: implicit declaration of function 'dewarpBuildPageModel_ex' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
return(dewarpBuildPageModel_ex(dew,debugfile,2));
       ^
dewarp2.c:144:8: note: did you mean 'dewarpBuildPageModel'?
dewarp2.c:141:1: note: 'dewarpBuildPageModel' declared here
dewarpBuildPageModel(L_DEWARP    *dew,
^
dewarp2.c:230:9: error: implicit declaration of function 'dewarpFindVertDisparity_ex' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
    if (dewarpFindVertDisparity_ex(dew, ptaa2, 0, fit_order) != 0) {
        ^
dewarp2.c:230:9: note: did you mean 'dewarpFindVertDisparity'?
./allheaders.h:654:25: note: 'dewarpFindVertDisparity' declared here
LEPT_DLL extern l_int32 dewarpFindVertDisparity ( L_DEWARP *dew, PTAA *ptaa, l_int32 rotflag );
                        ^
dewarp2.c:301:8: error: implicit declaration of function 'dewarpFindVertDisparity_ex' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
return(dewarpFindVertDisparity_ex(dew,ptaa,rotflag,2));
       ^
dewarp2.c:1411:8: error: implicit declaration of function 'dewarpBuildLineModel_ex' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
return(dewarpBuildLineModel_ex(dew,opensize,debugfile,2));
       ^
4 errors generated.
make[7]: *** [dewarp2.lo] Error 1
make[7]: *** Waiting for unfinished jobs....
make[6]: *** [install-recursive] Error 1
make[5]: *** [liblept.dylib] Error 2
make[4]: *** [libk2pdfopt-prefix/src/libk2pdfopt-stamp/libk2pdfopt-build] Error 2
make[3]: *** [CMakeFiles/libk2pdfopt.dir/all] Error 2
make[2]: *** [all] Error 2
make[1]: *** [build/x86_64-apple-darwin19.6.0/libs/libk2pdfopt.2.dylib] Error 2
make: *** [base/build/x86_64-apple-darwin19.6.0/luajit] Error 2

#### Steps to reproduce

git clone https://github.com/koreader/koreader.git
cd koreader
export PATH="/usr/local/bin:/usr/local/sbin:${PATH/:\/usr\/local\/bin/}"\n
export MACOSX_DEPLOYMENT_TARGET=10.14
./kodev fetch-thirdparty
./kodev release macos

**Comment by NiLuJe** [comment]
Will hopefully magically get fixed once https://github.com/koreader/koreader-base/pull/762 & co gets taken care of.

(i.e., not soon).

In the meantime, I'll choose to blame XCode 12 ;p.

**Comment by NiLuJe** [comment]
Or, better yet: why the hell is `-Werror` enabled here?

**Comment by JasonInOttawa** [comment]
The following note is in the README for base/thirdparty/leptonica/build/git_checkout/leptonica

   A brief version chronology is maintained in version-notes.html.
   Starting with gcc 4.3.3, error warnings (-Werror) are given for
   minor infractions like not checking return values of built-in C
   functions.  I have attempted to eliminate these warnings.
   In any event, you will see warnings with the -Wall flag.

It seems gcc 4.3.3 was released in 2009. The version shown by CC in XCode 12 is:

 % CC --version
Apple clang version 12.0.0 (clang-1200.0.32.2)

Perhaps it's taken them this long to catch up :) and finally show errors on these minor infractions.

Do you happen to have a pointer on how I could disable/override the -Werror flag, and any idea whether this would affect things badly?

Thank you very much!

**Comment by NiLuJe** [comment]
The only -Werror flags I can see in a standalone checkout of the same leptonica version are commented out?

(I don't have an actual KOReader checkout to double-check right now).

But if you figure out where it's coming from, it's completely harmless to patch it out, yeah.

**Comment by JasonInOttawa** [comment]
They all look commented out to me too. I wanted to try adding a "-w" flag (which I think means to turn off all the diagnostics, which sounds a bit scary) but couldn't figure out where to add it.

Moving the required functions to above where they are implicitly declared actually does seem to "fix" the errors. Old sloppiness hits new warning/error tightening. It probably is XCode 12's "fault". :) I see lots of other clang warnings during the thirdparty builds that don't halt things.

These errors also pop up in minizip:

entropy.c:44:16: error: implicit declaration of function 'read' is invalid in C99 [-Werror,-Wimplicit-entropy.c:45:9: error: implicit declaration of function 'close' is invalid in C99 [-Werror,-Wimplicit-2 errors generated.

Adding function declarations fixed those.

Hopefully closing the issue you noted above will fix things properly and let the build proceed with -Werror. Any idea what "not any time soon" means? :)

[entropy.c.txt](https://github.com/koreader/koreader/files/5251976/entropy.c.txt)
[dewarp2.c.txt](https://github.com/koreader/koreader/files/5251986/dewarp2.c.txt)

**Comment by NiLuJe** [comment]
If no-one volunteers to save me from having to do it, Q1 2021, I won't have the time to deal with it earlier.

**Comment by NiLuJe** [comment]
So, yeah, if you can just patch the individual things to be C99-friendly, don't hesitate to PR that in the meantime ;).

**Comment by Frenzie** [comment]
I'm moving next month, so I probably won't be swimming in time for at least several weeks. ;-)

**Comment by JasonInOttawa** [comment]
Is it possible to patch those files? I thought they got downloaded at build time. Or could they be copied in a script just prior to building.

**Comment by Frenzie** [comment]
It is, but we might also consider that we directly control https://github.com/koreader/libk2pdfopt That being said, a patch file here might be preferable wrt upstream k2pdfopt?

See the various patches in https://github.com/koreader/koreader-base/tree/7c380d0590ce51795513c3fc12bc6f36a2d9a131/thirdparty/mupdf

https://github.com/koreader/koreader-base/blob/7c380d0590ce51795513c3fc12bc6f36a2d9a131/thirdparty/mupdf/CMakeLists.txt#L68-L85

https://github.com/koreader/koreader-base/blob/7c380d0590ce51795513c3fc12bc6f36a2d9a131/thirdparty/mupdf/CMakeLists.txt#L102

**Comment by JasonInOttawa** [comment]
Thank you - that is very helpful. I will also look into adding a flag to the makefile to have it stop throwing errors on implicit declarations, in k2pdfopt and minizip.

**Comment by Frenzie** [comment]
If it's just the Makefile please PR it in libk2pdfopt. :-)

**Comment by JasonInOttawa** [comment]
I'm looking into producing patch files for k2pdfopt and minizip, to change the following files:

for dewarp2.c (libk2pdfopt)

139a140,141
> l_int32 dewarpBuildPageModel_ex(L_DEWARP *dew, const char *debugfile, l_int32 fit_order);
>
146a149,150
> l_int32 dewarpFindVertDisparity_ex(L_DEWARP *dew, PTAA *ptaa, l_int32 rotflag, l_int32 fit_order);
>
1405a1410,1411
> l_int32 dewarpBuildLineModel_ex(L_DEWARP *dew, l_int32 opensize, const char  *debugfile, l_int32 fit_order);
>

for entropy.c (minizip)
37a38,40
> int read(int frand, unsigned char buf[], unsigned int len);
> void close(int frand);
> 

Does this sound like a good way to proceed? I'd appreciate any guidance you could provide on how to produce the patch files. Thank you for pointing at mupdf.

**Comment by NiLuJe** [comment]
`git diff`?

Not *quite* sure what you're asking, exactly?

**Comment by JasonInOttawa** [comment]
Sorry for the confusion. The patch files in mupdf have what looks like diff output. The code snippets posted above are the difference between what git pull provides and source files modified with explicit function declarations to avoid the implicit declaration errors during compile.

I don't know which tools to use to create something that looks like the patch files in mupdf.

**Comment by NiLuJe** [comment]
* Start from a clean git clone of what you want to patch
* checkout the same version that we build (check thirdparty/<project>/CMakeLists.txt (or whateder CMake's terrible standard project filename actually is).
* Modify code with your favorite editror
* git diff > filename.patch

**Comment by Frenzie** [comment]
If you commit it you can also use `git format-patch <hash>`.

**Comment by JasonInOttawa** [comment]
Thanks very much. I'm trying to figure out how to create a pull request now :)

Attached are proposed changes in libk2pdfopt and minizip.
[minizip_CMakeLists.txt](https://github.com/koreader/koreader/files/5262146/minizip_CMakeLists.txt)
[minizip_add_explicit_fn_decl.patch.txt](https://github.com/koreader/koreader/files/5262147/minizip_add_explicit_fn_decl.patch.txt)
[libk2pdfopt_add_explicit_fn_decl.patch.txt](https://github.com/koreader/koreader/files/5262156/libk2pdfopt_add_explicit_fn_decl.patch.txt)
[libk2pdfopt_CMakeLists.txt](https://github.com/koreader/koreader/files/5262157/libk2pdfopt_CMakeLists.txt)

**Comment by JasonInOttawa** [comment]
I've got change lined up as follows:

git status
HEAD detached at 0092277
Changes to be committed:
  (use "git restore --staged <file>..." to unstage)
	modified:   thirdparty/libk2pdfopt/CMakeLists.txt
	new file:   thirdparty/libk2pdfopt/add_explicit_fn_decl.patch
	modified:   thirdparty/minizip/CMakeLists.txt
	new file:   thirdparty/minizip/add_explicit_fn_decl.patch

I've been trying to following the instructions at:

http://koreader.rocks/doc/topics/Collaborating_with_Git.md.html

as follows:

jasonbenwell@Jasons-MacBook base % git remote add upstream git@github.com:koreader/koreader-base.git

jasonbenwell@Jasons-MacBook base % git remote -v show

origin	https://github.com/koreader/koreader-base.git (fetch)
origin	https://github.com/koreader/koreader-base.git (push)
upstream	git@github.com:koreader/koreader-base.git (fetch)
upstream	git@github.com:koreader/koreader-base.git (push)

But I can't push my changes:

jasonbenwell@Jasons-MacBook base % git remote -v
origin	https://github.com/koreader/koreader-base.git (fetch)
origin	https://github.com/koreader/koreader-base.git (push)
upstream	git@github.com:koreader/koreader-base.git (fetch)
upstream	git@github.com:koreader/koreader-base.git (push)
jasonbenwell@Jasons-MacBook base % git push upstream jasoninottawa-add_explicit_fn_decl
git@github.com: Permission denied (publickey).
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.

I'm quite certain I'm doing this entirely wrong. If there is any other documentation on how to create a PR, I'd appreciate it if you could point me to it.

**Comment by poire-z** [comment]
You should push to "origin" (where you have the rights, "upstream" is KOReader's where you don't)
Not sure about your state of things with `HEAD detached at 0092277`.
I expect you should be on a new named branch to make a PR.

Keep a copy of your modified file, and may be clean it up and follow the instructions I wrote recently at
https://github.com/koreader/crengine/pull/378#issuecomment-694169843 (with `koreader-base` instead of `crengine`).

**Comment by Frenzie** [comment]
Wait, I'm a bit confused, why didn't you just patch out -Werror? :-)

**Comment by Frenzie** [comment]
Also even though it's harmless on other platforms it might be best to stick an `if MACOS` (or do we use `DARWIN`) around it to make it self-documenting why it's there.

**Comment by JasonInOttawa** [comment]
I tried finding where it was coming from and couldn't :) They all seem to be commented out. I can look again.

**Comment by JasonInOttawa** [comment]
Thank you poire-z for the pointers

**Comment by JasonInOttawa** [comment]
The attached add

if ($ENV{DARWIN})
    #fix build error due to implicit declarations of function being invalid in C99 under macOS/XCode 12
    set(CFLAGS "${CFLAGS} -Wno-error=implicit-function-declaration")
endif()

to libk2pdfopt and minizip, which allows the build to complete without turning off too much error checking.

Minizip also needs the CFLAGS added to BUILD_CMD_STR:

if ($ENV{DARWIN})
    set(BUILD_CMD_STR -j${PARALLEL_JOBS} -C aes CC=${CC} CFLAGS=${CFLAGS} LDFLAGS=${LDFLAGS} AR=${AR} RANLIB=${RANLIB})
else()
    set(BUILD_CMD_STR -j${PARALLEL_JOBS} -C aes CC=${CC} LDFLAGS=${LDFLAGS} AR=${AR} RANLIB=${RANLIB})
endif()

If this is better, I'll look into putting it into a PR.
[minizip-CMakeLists.txt](https://github.com/koreader/koreader/files/5263961/minizip-CMakeLists.txt)
[libk2pdfopt-CMakeLists.txt](https://github.com/koreader/koreader/files/5263962/libk2pdfopt-CMakeLists.txt)

**Comment by JasonInOttawa** [comment]
I'm having a little bit of trouble at the git pull. (I do hope I'm not asking too many questions)

jason@Jasons-MacBook base % git push -u origin jasoninottawa-allowimplicitfndecl
Username for 'https://github.com': JasonInOttawa
Password for 'https://JasonInOttawa@github.com': 
remote: Permission to koreader/koreader-base.git denied to JasonInOttawa.
fatal: unable to access 'https://github.com/koreader/koreader-base.git/': The requested URL returned error: 403
jason@Jasons-MacBook base % git branch
* (HEAD detached from 2ef3125)
  jasoninottawa-allowimplicitfndecl
  master
jason@Jasons-MacBook base % git remote -v
origin	https://github.com/koreader/koreader-base.git (fetch)
origin	https://github.com/koreader/koreader-base.git (push)
upstream	git@github.com:koreader/koreader-base.git (fetch)
upstream	git@github.com:koreader/koreader-base.git (push)

**Comment by poire-z** [comment]
Your "origin" is wrong: it is koreader's koreader. It should be https://github.com/JasonInOttawa/koreader (which doesn't seem to exist, you have not forked the repository, it's not visible in https://github.com/JasonInOttawa?tab=repositories).
Re-start from scratch: keep a backup of your modified files, remove your local close, follow my instructions from the link above from the start.
FORK repo on github with: ![image](https://user-images.githubusercontent.com/24273478/93972638-f105a700-fd72-11ea-97b9-ba21fbc0cc35.png)
Check you have koreader-base in your repos.
Then, on your computer:
```
$ git clone https://github.com/JasonInOttawa/koreader-base.git
$ cd koreader-base
$ git remote -v     should show:
    origin  https://github.com/JasonInOttawa/koreader-base.git (fetch)
    origin  https://github.com/JasonInOttawa/koreader-base.git (push)
    upstream  https://github.com/koreader/koreader-base.git (fetch)
    upstream  https://github.com/koreader/koreader-base.git (push)
```

**Comment by JasonInOttawa** [comment]
Thanks very much for all your help and patience! I've opened PR #1198.

With regard to doing a complete build of Koreader, is there a git clone command which will pull the master koreader and my fork of koreader-base into the "base" directory?

**Comment by Frenzie** [comment]
I'm not quite sure what you mean, but a git submodule is pretty much just a regular repository. You can add different remotes, check them out, etc.

But with a pull request on GH there's another little trick you can use to checkout the PR code:
```sh
# or upstream instead of origin
git fetch origin pull/1198/head
git checkout FETCH_HEAD
```

That's how I test PRs most of the time; I only do branches/remotes if I need to push something more complex and that's seldom the case outside of my own PRs.

**Comment by NiLuJe** [comment]
> I'm not quite sure what you mean, but a git submodule is pretty much just a regular repository. You can add different remotes, check them out, etc.
> 
> But with a pull request on GH there's another little trick you can use to checkout the PR code:
> 
> ```shell
> # or upstream instead of origin
> git fetch origin pull/1198/head
> git checkout FETCH_HEAD
> ```
> 
> That's how I test PRs most of the time; I only do branches/remotes if I need to push something more complex and that's seldom the case outside of my own PRs.

The new [gh cli](https://github.com/cli/cli) can also help hide the complexity away behind simpler commands ;).

**Comment by JasonInOttawa** [comment]
Sorry, another question: I've changed some things since I created the PR and I'd like to push them to the repository. 

I've done git adds:

jason@Jasons-MacBook koreader-base % git status
On branch add_allow_impl_fn_decl
Your branch is up to date with 'origin/add_allow_impl_fn_decl'.

Changes to be committed:
  (use "git restore --staged <file>..." to unstage)
	modified:   thirdparty/libk2pdfopt/CMakeLists.txt
	modified:   thirdparty/minizip/CMakeLists.txt

I'm trying to following this step:

Add more commits (as above) if needed to answer review comments, and push them:
jason@Jasons-MacBook koreader-base % git push origin add_allow_impl_fn_decl
Everything up-to-date

git status shows this

jason@Jasons-MacBook koreader-base % git status
On branch add_allow_impl_fn_decl
Your branch is up to date with 'origin/add_allow_impl_fn_decl'.

I'm assuming there's a step I should be doing before git push, but I'm not sure what it is :)

**Comment by JasonInOttawa** [comment]
Meant to include this

jasonbenwell@Jasons-MacBook koreader-base % git status
On branch add_allow_impl_fn_decl
Your branch is up to date with 'origin/add_allow_impl_fn_decl'.

Changes to be committed:
  (use "git restore --staged <file>..." to unstage)
	modified:   thirdparty/libk2pdfopt/CMakeLists.txt
	modified:   thirdparty/minizip/CMakeLists.txt

**Comment by poire-z** [comment]
> I've changed some things since I created the PR

But `git status` tells you there's something new (which should be red)
```
Changes to be committed:
(use "git restore --staged ..." to unstage)
modified: thirdparty/libk2pdfopt/CMakeLists.txt
modified: thirdparty/minizip/CMakeLists.txt
```
So, add them and create a commit:
```
git add thirdparty/libk2pdfopt/CMakeLists.txt thirdparty/minizip/CMakeLists.txt
git status     should show them in green
git commit -v
```
and then, `git push` should work

Please wrap your logs or terminal output after you paste it, to make it easier to read for us:
![image](https://user-images.githubusercontent.com/24273478/94034020-d0187280-fdc1-11ea-900f-f1cfa000a50a.png)

**Comment by JasonInOttawa** [comment]
Thank you! That

`WORKED!`

:)

**Comment by harrylepotter** [comment]
Hey @JasonInOttawa  did this get merged? Seem to be having the same issue on Catalina.
I notice the PR dates back to 2014?

**Comment by JasonInOttawa** [comment]
The PR mentioned is actually in koreader-base:

`https://github.com/koreader/koreader-base/pull/1198`

It isn't merged yet but if you download the revised files and put them in the respective spots:

```
base/thirdparty/libk2pdfopt/CMakeLists.txt
base/thirdparty/minizip/CMakeLists.txt
```

your build should complete.

I think the issue was introduced by XCode 12 - are you running that? I think that prior to XCode 12, you could have implicit function declarations, but that got tightened up in XCode 12. The change puts in a flag that allows implicit function declarations.

**Comment by JasonInOttawa** [comment]
I'm happy with the fix in koreader-base

Fix build in macOS XCode 12 #1198
